storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -88,85 +488,214 @@ type ReconcileVolume struct {
-	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "", Name: pv.Name}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new PersistentVolume", "PersistentVolume.Name", pv.Name)
-		err = r.client.Create(context.TODO(), pv)
+	switch volume.Status.Job {
+	case "": // No job in progress: call Salt to prepare the volume.
+		if jid, err := self.salt.PrepareVolume(ctx, nodeName); err != nil {
+			reqLogger.Error(err, "failed to run PrepareVolume")
+			return delayedRequeue(err)
+		} else {
+			reqLogger.Info("start to prepare the volume")
+			self.recorder.Event(
+				volume, corev1.EventTypeNormal, "SaltCall",
+				"Salt call to PrepareVolume started",
+			)
+			return self.setPendingVolumeStatus(ctx, volume, jid)
+		}
+	case JOB_DONE_MARKER: // Salt job is done, now let's create the PV.
+		diskSize, err := self.getDiskSizeForVolume(ctx, volume)
 		if err != nil {
-			return reconcile.Result{}, err
+			reqLogger.Error(err, "call to GetVolumeSize failed")
+			return self.setFailedVolumeStatus(
+				ctx, volume, storagev1alpha1.CreationError,
+				"call to GetVolumeSize failed: %s", err.Error(),
+			)
+		}
+		pv := newPersistentVolume(volume, diskSize)
+		if err := self.createPersistentVolume(ctx, pv, volume); err != nil {
+			return delayedRequeue(err)
 		}
+		reqLogger.Info(
+			"creating a new PersistentVolume", "PersistentVolume.Name", pv.Name,
+		)
+		self.recorder.Event(
+			volume, corev1.EventTypeNormal, "PvCreation",
+			"backing PersistentVolume created",
+		)
+		return self.setAvailableVolumeStatus(ctx, volume)
+	default: // PrepareVolume in progress: poll its state.
+		return self.pollSaltJob(
+			ctx, "PrepareVolume", volume, self.setPendingVolumeStatus,
+		)
+	}
+}
 
-		// PV created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
-		return reconcile.Result{}, err
+// Finalize a volume marked for deletion.
+func (self *ReconcileVolume) finalizeVolume(
+	ctx context.Context,
+	request reconcile.Request,
+	volume *storagev1alpha1.Volume,
+) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Name", request.Name)
+
+	if volume.Status.Phase == storagev1alpha1.VolumePending {
+		reqLogger.Info("pending volume cannot be marked for deletion: requeue")
+		return delayedRequeue(nil)
 	}
 
-	// PV already exists - don't requeue
-	reqLogger.Info("Skip reconcile: PersistentVolume already exists", "PersistentVolume.Name", found.Name)
+	// Check if a PV is associated to the volume.
+	pv, err := self.getPersistentVolume(ctx, volume.Name)
+	if err != nil {
+		reqLogger.Error(
+			err, "cannot read PersistentVolume: requeue",
+			"PersistentVolume.Name", volume.Name,
+		)
+		return delayedRequeue(err)
+	}
 
-	return reconcile.Result{}, nil
+	// No associated PV: let's delete ourselves then.
+	if pv == nil {
+		reqLogger.Info("no PersistentVolume associated to terminating volume")
+		if err := self.removeVolumeFinalizer(ctx, volume); err != nil {
```

Comment:
Then where do we 'delete' the volume (through Salt)?

Reply:
Right, the Salt call to unprovision is done when the PV still exists, and we rely on the PV to exist to trigger it. Which may be incorrect, if the Volume is DELETEd before we create the PV, decide not to create it, and would end up here. I'm afraid the current approach is a little too brittle, and inconsistent:

- We provision the volume through Salt *before* creating a PV
- Now suddenly, we would unprovision such volume through Salt before *deleting* the PV

Reply:
Don't get me wrong, deleting the PV after the volume has been unprovisioned (through Salt) is likely the right approach, though I don't think we should consider (non-)existence of a PV as a marker.

Reply:
Finalization flow was reworked to address this aspect (I think…)